### ISSUE TITLE
Frontend work to implement multiple reactions (V1)

### ISSUE THREAD

**Issue by mirie** [forem/forem#18641]
## User Story (Who / What / Why)

As a user, I want the ability to leave multiple reactions on posts in order to have more options to express myself and engage with content on the platform.

As an admin, I want to choose the specific reaction options on my community to ensure they resonate with my community members.

## Additional Context

We are going to take on this project in different phases:

- V1: The Forem must use defined set of hardcoded reactions
- V2: Admins can choose from a few different "sets" of predefined reactions (Is handled as a followup issue in #18703)
- V3: Admins can configure any reaction by uploading icons, a name, etc. (similar to Navigation Links)

_*Helpful note from our refinement discussion: Split the post view and homepage work into two separate PRs for easier testing/review/reverts ❤_ 

## Definition of Done for V1

Core Team can access the Figma file [here](https://www.figma.com/file/Rl50gxO5BHioRC3nUoRQ7M/Reactions-%E2%9D%A4%EF%B8%8F?node-id=988%3A3192).  We are using V1.1

**Post Index View:** (When a post card shows up in a feed, profile page, etc.)
- [ ] Display # of reactions associated with the article (consistent with old logic of showing the net amount of Hearts + Unicorns)

**Post View:**
- [ ] Display # of reactions associated with the article (consistent with old logic of showing the net amount of Hearts + Unicorns)
- [ ] Add ability for the user to create their own reactions per Figma specs
- [ ] Add reactions and totals above title (below author) per Figma specs
    - [ ] Add ability for user to add / remove their own reaction by interacting with this horizontal row of reactions (stretch?)

**Cleanup existing feature flags/finalize last bits previous work with reactions**
- [ ] Cleanup feature flags as detailed in https://github.com/forem/forem/issues/18277#issuecomment-1265663831 and close that open issue. -- linked draft PR with this work: https://github.com/forem/forem/pull/18410 

## QA and Coordination (for rollout):
- [ ] Make a note in #product-management that the feature is ready to be shipped so that we can coordinate final QA
- [ ] @anujbhavsar96 to participate in Design QA for this so loop in so we can QA on either Canary 1 or 2

## Teaser Screenshots

![Reactions Homepage](https://user-images.githubusercontent.com/9434801/199347577-bf5db7ec-e8e0-4821-82a4-8441f39790d9.png)

![Reactions Granular](https://user-images.githubusercontent.com/9434801/199347495-3419b3d1-c5ab-49b0-8065-2a4035a2b110.png)

**Comment by rt4914** [comment]
The scope and definition of V1 looks clear and correct to me.

**Comment by jaw6** [comment]
> Add ability for user to add / remove their own reaction by interacting with this horizontal row of reactions (stretch?)

The reactions system is more of a toggle than a `create` -- a second click removes the previous reaction of same type -- I believe it would likely be less work to preserve this as-is than change it.

Some questions:

* What are the hard-coded reactions for V1? In the Figma, it looks like we might have `like` ❤️ , `bulb` 💡, `rocket` 🚀 , `raised_hands` 🙌 , and `tada` 🎉 . Are those the intended text labels, as of V1?
* How (or maybe when) are we addressing accessibility impact with this feature? (ie, it seems we'll need to add a human label for new reactions? and check that the new reaction pop-up is screenreader-friendly?)
* `unicorn` 🦄 reactions are currently in a weird status -- jump-to-comments is feature-flagged, but 🦄 -code remains in place. Should V1 include 🦄 cleanup? Is jump-to-comment staying as-is? In #18521, there was a little discussion of having categories be removable/unpublishable -- should we be treating 🦄 as a published or unpublished reaction category? (Or, should we [be coercing 🦄 to ❤️ ?](https://github.com/forem/forem/pull/18410) ??)
* Not mentioned here is the point-scoring mechanism underlying reactions -- previously, I think we were assuming/treating all reactions as worth `1.0` point. Is the plan to continue that scoring throughout these three iterations?
* The reactions API isn't included in the definition of done for any version here -- that API's primary customer is anti-spam `forem-shield`, but presumably it should also be updated to make use of the new reaction categories at some point?

`/admin/content_manager/articles`
<img width="1006" alt="Screen Shot 2022-11-04 at 11 31 02" src="https://user-images.githubusercontent.com/2077/199951901-88322ef3-7fac-4fef-b7bc-4da802b01250.png">

Should this view in Admin show details for reaction categories other than `like` ❤️  ?

**Comment by benhalpern** [comment]
Hey @jaw6, I'll circle back up on Monday if any of this needs to be addressed with more detail or nuance, but here are the answers as I see it:

- We will finalize the new hardcoded reactions within a few days.
- The label should be accounted for with the "name" feature we're giving to all reactions, I.e. "Reading List", and that can be used where needed on the frontend. If this is an incomplete answer, let's talk about it.
- I agree that between the unicorn work sort of being in limbo and the scenario of "reintroduction" could present some problems. I think we'll need to address this specifically as a question of tactics and _discuss_ edge cases of initial rollout. We _will_ be keeping unicorn around as a default hardcoded one upon rollout, so keep that in mind. I hope any problems are mostly just initial rollout and not long term hitches.
- Public reactions will default to 1.0 and stay that way through the current launch iteration. It is possible that this could become a future consideration. If we "future proof" by adding a `points` column now, we may not end up ever using it. Curious how you feel.
- Yes, as you mentioned — anti-abuse via `forem-shield` is the main use case here, but I'd imagine for systematic integrity, we should likely ensure that it remains functional with each step in the iteration process.

**Comment by mirie** [comment]
@anujbhavsar96 a request from our refinement meeting: Can you delineate V1 from V2 in the figma files and I'll update both issues afterwards. Thanks!

**Comment by anujbhavsar96** [comment]
> Add ability for user to add / remove their own reaction by interacting with this horizontal row of reactions (stretch?)

@mirie Users can only react from the reaction drawer only. This reaction information above the title is to display engagement upfront. And going forward, we will use this information to display [who reacted with a particular reaction](https://github.com/forem/forem/issues/18125) in the modal on click.

**Comment by mirie** [comment]
@jaw6 I know that you've been working on a diagram -- can you attach to this issue?

**Comment by jaw6** [comment]
Repeating comment here from the PR:

Status of the PR:

* [x] I have a YML file with configurable reaction categories.
* [x] That file controls the display and order of possible reaction buttons on the `#show` page.
* [x] The buttons are ugly -- they are grey-scale and not in a drawer -- but functional (clicking them works to reaction as expected, counts update, etc.)

<img width="96" alt="Screen Shot 2022-12-13 at 16 24 25" src="https://user-images.githubusercontent.com/2077/207380511-33346eeb-f95f-411a-8128-c39e0662b875.png">

What needs doing:

* [ ] Aggregate reaction counts on `#index` -- I'm working on that now, it should be straight-forward.
* [ ] I need to figure out how to deal with `readinglist` being a `reaction category` that _isn't_ in the drawer. A simple fix is to add another attribute to the YML for this, but I suspect this won't be the last time we have to have special-case behaviour for this weird reaction type. (For example: does it add to reaction counts on `#index` -- I don't think it does, if not, that's another area to special-case...)
* [ ] Color: The heart/unicorn/bookmark icons are rendered as monochrome SVGs and are given color (when active or hovered) via CSS that incorporates the reaction category. In other words, [`like` has its own CSS color](https://github.com/forem/forem/blob/main/app/assets/stylesheets/views/article.scss#L211-L233) (and it only gets this color when it's active or hovered). The [figma designs](https://www.figma.com/file/Rl50gxO5BHioRC3nUoRQ7M/Reactions-%E2%9D%A4%EF%B8%8F?node-id=988%3A3192&t=mJxLk7ZP2E492uJa-0) for this have always-full-color icons and colorized hover/active background/border. I don't see these exact icons in twemoji's SVG assets and I'm trying to think how to get variable `hover` or `active` color out of the YML and into the CSS, but a simplified initial step might be to have the `active` or `hover` states for every reaction use the same color?
* [ ] An actual drawer effect - I have been leaving this until I got the rendering under control (no need for an extra reveal step when I'm refreshing this much), but I could use a hand with this, actually, front-end drawers are not my strong suit.
* [ ] The issue requests that we deal with unicorn cleanup as _part_ of this roll-out, however I think we should actually deal with unicorns _after_ roll-out, as a part of feature-flag cleanup post-deploy. (We're going to be rolling multiple-reactions out under a feature flag. unicorns are currently feature-flagged to toggle with jump-to-comments. When multiple-reactions is ready to be un-flagged, then unicorns can transition to being un-published reactions.)

I want to call out a small detail concerning **I18n and ARIA**: One impact of this change will be making the site _slightly less_ translation-friendly. In the "old" mode, reactions are part of the site UI. We use Rails' I18n mechanisms to label these buttons (though our "french" locale files are all in english, for now). There are additional labels (generated out of I18n) that feed additional accessibility information to ARIA. In the "new" mode, reactions are part of site **contents** and are not translated. We can provide a `label` hook (and even an `aria-label` hook) in the YML files (and eventually when this moves to the DB) but -- as far as I know -- we don't have an I18n mechanism set up that hooks into "site content".

**Comment by mirie** [comment]
Thank you for this detailed update @jaw6!

@anujbhavsar96 can you take a look at the "Color" bullet and let us know your thoughts on the compromise proposed?

@jaw6 sounds like the drawer effect can be broken out to be worked on by another engineer. Does it make sense to try to start on this now (branching off your branch) or wait because there's other pieces still in progress?

cc @benhalpern re: some of the questions/concerns called out in the last comment

**Comment by benhalpern** [comment]
@jaw6 on the i18n, I think we are going to have to move forward with not fully handled yet  — acknowledging the yet incomplete status of i18n in general. This is one of a number of areas where content provided by admins bleeds into UI, and it will be something we will have to address to consider i18n stable.

I think that part of things will be on the table as a component of our Q1 product stabilization efforts, so I think we can move forward without that being handled yet, and make note of it.

I think @anujbhavsar96 should clarify the expectations we landed on with color, and the status of icons. There might be some misalignment here.

**Comment by anujbhavsar96** [comment]
@mirie and @jaw6 One thing to notice is in the current implementation, we have two types of icons for the normal(stroke icon) and active(filled icon) state. 

With this new reaction mechanism, we wanted to make it a visually appealing and fun experience. That’s the only reason we decided to keep the normal and active state as colored icons (Similar to FB) because current flat-style monochrome icons look boring. Another reason to keep a single colored icon for all states is in V3 _(where admins can configure any reaction by uploading icons, a name, etc.)_ we will be asking for admin inputs for configurations; at that time, we want to minimize admin inputs as much as possible to make the UX flawless. If we keep the mechanism of two types of icons like the current one, going forward, we will have to ask for two kinds of icons from admins one for the default/hover state(monochrome icon) and another for the active state(colored icon). 

Also, these icons are not available in the Twemoji library. These are custom icons.

Right now, hovering the Heart monochrome icon will highlight both the icon and background. But in new designs, we are highlighting the background only since icons will be colored icons already. Going forward, we can experiment here to make the hover state more prominent and fun where the icon moves up or wiggles on the hover.

So IMO, if we can make this work, this would be a huge win for the upcoming iterations. Otherwise, below is the approach we can take, which is similar to the current one where we display monochrome stroke icons as default/inactive state, a colored version of that icon as hover, and the final multicolor icon as the active state. cc @benhalpern 

![Dec-14-2022 10-22-01](https://user-images.githubusercontent.com/84626054/207510262-2e7ac256-d3c6-4f92-bfc1-645ce3da440a.gif)

Let me know what you all think ✨

**Comment by jaw6** [comment]
Status of various things:

* [ ] Aggregate reaction counts on `#index` -- not as straight-forward as I had hoped, actually. What I did not know before this week is that logged-out users get an `#index` rendered server-side and logged-in users get an `#index` rendered client-side. The problem with this for reactions is that -- as far as I know -- we don't have hooks for `FeatureFlag` client-side, so I'm not sure we have a mechanism to tell the client to render a different template. If we're _okay_ with the `#index` changes only being visible to logged-out users, though...
* [ ] Color: I figured out how crayons renders full-color icons, but I still don't see a mechanism to render custom per-reaction backgrounds. @anujbhavsar96 can we pick one highlight color maybe?
* [x] "readinglist" is special: I think I've got this working, but it's just a bit ugly. I don't see a better solution, but maybe something will come out of PR review.
* [x] "Pseudo drawer": I made a prototype "drawer" effect, using crayons' `dropdown` widget. It opens with a click instead of a hover, so that's not ideal, and I'm not confident it'll work well on mobile, but it's a starting point maybe?
* [x] i18n & Aria - sounds like we're okay with this rolling out with less i18n & aria
* [ ] unicorn - are we okay with unicorn cleanup happening as part of feature-flag cleanup?

`index (logged-out)`
<img width="678" alt="Screen Shot 2022-12-16 at 10 37 27" src="https://user-images.githubusercontent.com/2077/208124560-2b26b0f5-1baa-4712-bf3b-8babd9061a4e.png">

`index (logged-in)`
<img width="670" alt="Screen Shot 2022-12-16 at 10 38 04" src="https://user-images.githubusercontent.com/2077/208124641-1f99ba17-52c3-4607-938b-672316a5f77e.png">

`pseudo-drawer` 
<img width="900" alt="Screen Shot 2022-12-16 at 15 43 41" src="https://user-images.githubusercontent.com/2077/208124725-3a77e444-529d-42bd-9b05-0ab11d1f9cbc.png">

(Things get a bit busy with the pseudo-drawer, tooltips, etc.)

**Comment by anujbhavsar96** [comment]
> Aggregate reaction counts on `#index`

@jaw6 If we display these icons on the index page, they should be visible for logged-in and logged-out users both. Otherwise, we can skip this guess. What do you think, @benhalpern ?

> Color: I figured out how crayons renders full-color icons, but I still don't see a mechanism to render custom per-reaction backgrounds. @anujbhavsar96 can we pick one highlight color maybe?

We can do that. We can use this kind of neutral color highlights, so it doesn't mess with brand color. What's your take on this @benhalpern ? 

<img width="722" alt="image" src="https://user-images.githubusercontent.com/84626054/208390550-845c2f79-c8a1-4848-8e56-5189e0e3fd32.png">

> Things get a bit busy with the pseudo-drawer, tooltips, etc.

We should probably open the drawer a  bit right side, so it doesn't overlap with icons on the action bar. Something like below

<img width="380" alt="image" src="https://user-images.githubusercontent.com/84626054/208389926-89103858-f589-41e1-85e5-52d454f31456.png">

**Comment by jaw6** [comment]
Current status:

* Grey color for user-activated
* I still have the multi-reaction badges on `index` for logged-out users (via feature-flag), but we can turn that off if we decide to focus on `show`.

<img width="333" alt="Screen Shot 2022-12-20 at 11 40 51" src="https://user-images.githubusercontent.com/2077/208674087-8ed6876e-4fa7-4dce-a99d-0cbc67dff730.png">

**Comment by mirie** [comment]
@maestromac Can we get this deployed to Canary 1 please so we can start playing around with it on a site?
@jaw6 as discussed during our 1:1 today, you want to get feedback on the PR in parallel (https://github.com/forem/forem/pull/18808) -- can you followup with tagged folks so that we can make sure to get timely feedback if you haven't gotten any by tomorrow. @benhalpern would love you to take a look at the PR as well in case you have feedback to share.


@benhalpern / @anujbhavsar96: Recapping a conversation I had with @jaw6 so we're all aligned.

- Having this deployed to Canary1 will give both of you the opportunity to play around the feature as currently implemented and note anything that is a **(go-live) priority**
- We can start QA-ing on Canary1 to note any issues not caught
- Regarding icons on the index page: @jaw6 I just realized that we cannot see pages on Canary1 as a logged out user. @maestromac is there a way we can? I would like to get early @anujbhavsar96 / @benhalpern feedback on the look of the article (at least for logged out users). I agree that for launch, we should NOT have a different experience for logged out / logged in users, so I propose rolling this part out after we're comfortable with cleaning out the feature flag for multiple reactions and having this experience available to everyone. It would be good to have feature flag support client side generally but want to separate that from at least the initial go live of multiple reaction functionality.

**Comment by maestromac** [comment]
Deployed to Canary1

**Comment by benhalpern** [comment]
@mirie adjusted Canary1 to be public (still requires @forem.com email for signing in)

**Comment by anujbhavsar96** [comment]
Sounds good. I'll play around with this feature on Canary1 and report any issues with documentation in Figma itself. I'll share it by Monday eod

**Comment by anujbhavsar96** [comment]
I've done the first pass of testing on Canary1. I've listed down all the issues in [Figma itself](https://www.figma.com/file/Rl50gxO5BHioRC3nUoRQ7M/Reactions-%E2%9D%A4%EF%B8%8F?node-id=1272%3A3141&t=71MohdIXHPPdaaCs-4). All the issue has some small description, how it looks right now in red outline and how it should look in green outline.

Let me know if anything looks confusing. I would be happy to explain it. Next, I'll test for Dark mode and mobile version

**Comment by anujbhavsar96** [comment]
I mentioned this in issues, but it requires a larger discussion about overall tooltip restructure.
<img width="638" alt="image" src="https://user-images.githubusercontent.com/84626054/213969779-608ef946-a00c-4e3a-bdd0-1ce2c2fd27d1.png">

But I agree with you, @benhalpern; we are probably better without a tooltip for now.

**Comment by jaw6** [comment]
I've opened https://github.com/forem/forem/issues/19008 to collect & track progress on the remaining issues with this feature, closing this issue in favor of that one.
